components/compliance-service/profiles/db/store.go

```diff
@@ -376,89 +379,108 @@ func (s *Store) Delete(namespace string, name string, version string) error {
+	return entries, total, nil
 }
 
-func (s *Store) ListProfilesMetadata(namespace string, name string, sort string, order string) ([]inspec.Metadata, error) {
-	if order != "ASC" && order != "DESC" {
-		return nil, status.Errorf(codes.InvalidArgument, "order field '%s' is invalid. Use either 'ASC' or 'DESC'", order)
-	}
-
-	if sort != "name" && sort != "title" && sort != "maintainer" {
-		return nil, status.Errorf(codes.InvalidArgument, "sort field '%s' is invalid. Use either 'name', 'title' or 'maintainer'", sort)
-	}
-
-	var rows *sql.Rows
-	var err error
-
-	switch {
-	case len(namespace) == 0 && len(name) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE sha256 IN
-			(SELECT sha256 FROM store_market)
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query)
-	case len(namespace) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE sha256 IN
-			(SELECT sha256 FROM store_market)
-			AND info->>'name' = $1
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, name)
-	case len(name) == 0:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace)
-	default:
-		query := fmt.Sprintf(`
-			SELECT sha256, info
-			FROM store_profiles WHERE
-			sha256 IN (SELECT sha256 FROM store_namespace WHERE owner=$1)
-			AND info->>'name' = $2
-			ORDER BY store_profiles.info #>> '{%s}' %s, store_profiles.info #>> '{version}' DESC;
-		`, sort, order)
-
-		rows, err = s.DB.Query(query, namespace, name)
+type ProfilesListRequest struct {
+	Filters   map[string][]string
+	Name      string
+	Namespace string
+	Order     string
+	Page      int
+	PerPage   int
+	Sort      string
+}
+
+func (s *Store) ListProfilesMetadata(req ProfilesListRequest) ([]inspec.Metadata, int, error) {
+	if req.Order != "ASC" && req.Order != "DESC" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "order field '%s' is invalid. Use either 'ASC' or 'DESC'", req.Order)
+	}
+
+	if req.Sort != "name" && req.Sort != "title" && req.Sort != "maintainer" {
+		return nil, 0, status.Errorf(codes.InvalidArgument, "sort field '%s' is invalid. Use either 'name', 'title' or 'maintainer'", req.Sort)
+	}
+
+	sql := squirrel.
+		Select("sha256, info, count(*) OVER() AS total").
+		From("store_profiles")
+
+	if len(req.Namespace) == 0 {
+		sql = sql.
+			Where("exists(select 1 from store_market where store_profiles.sha256 = store_market.sha256)")
+	} else {
+		sql = sql.
+			Where("exists(select 1 from store_namespace where store_profiles.sha256 = store_namespace.sha256 and store_namespace.owner = ?)", req.Namespace)
+	}
```

Comment:
The original hardcoded SQL did `where sha256 in (select sha256....` instead of the exists query. This is because the squirrel library's `Where` will do escape parameters for you but it won't accept arbitrary SQL fragments. The two queries should be equivalent and according to this issue on the squirrel bug tracker, the `exists` version should be faster in postgres: https://github.com/Masterminds/squirrel/issues/70

Reply:
I think the idiomatic way to do this would be to use a left outer join, but that seemed like a bigger change.

Reply:
Nice; thanks for the comment @phiggins !
You can take this further, though, if you want: instead of the hybrid (part Squirrel, part SQL) you could do nested squirrel--see https://github.com/Masterminds/squirrel/blob/7fcba632b69f0b1c1c60e71cce11bcbf6320a406/select_test.go#L203-L211 for the relevant use case.

Reply:
Just from a glance at that example it doesn't look like it gains us much over the sql fragment besides just consistency in always using the library as much as possible. I'll play around with it though, I thought I remembered seeing an `Exists` method (maybe only on master?).